### ISSUE TITLE
Both `_selected_range` and `_replacement_range` parameters are ignored in NSTextInputClient implementation.

### ISSUE THREAD

**Issue by ShikiSuen** [rust-windowing/winit#3617]
### Description

Both `_selected_range` and `_replacement_range` parameters are ignored in NSTextInputClient implementation. This causes an issue that these parameters sent from IME to IMKTextInput APIs are completely neglected:

https://github.com/rust-windowing/winit/blob/44aabdddcc9f720aec860c1f83c1041082c28560/src/platform_impl/macos/view.rs#L321

These parameters are crucial for some IMEs utilizing nested inline composition buffers (i.e. nested preedits). E.g. The Traditional Chinese Zhuyin IME shipped in macOS.

> Note: These parameters are UTF16 ranges. Apple introduced these APIs in macOS 10.5 Leopard.

I was about to send a PR to fix this issue. However, I have to figure out one thing:

In `src/events.rs`, search `pub enum Ime` and find `Preedit(String, Option<(usize, usize)>),`. I have doubts about this property: Are these cursor positions utf8 or utf16? This has to be figured out prior to patching this issue.

<img width="815" alt="image" src="https://github.com/rust-windowing/winit/assets/3164826/18786756-ae93-47c5-a023-17a748285c0c">

### macOS version

```shell
ProductName:		macOS
ProductVersion:		14.4.1
BuildVersion:		23E224
```

### Winit version

Master Branch 44aabddd

**Comment by kchibisov** [comment]
Everything is utf8, it's also documented like that.

**Comment by ShikiSuen** [comment]
@kchibisov Thanks. Could you please explain why the documentation (in the screenshot above) says "The cursor positioin is byte-wise indexed"?

**Comment by kchibisov** [comment]
It's an offset into utf8's byte string representation, because `String` is `utf8`.

**Comment by ShikiSuen** [comment]
@kchibisov So it is still not the utf8 literal char count?

**Comment by ShikiSuen** [comment]
It seems that currently available intels are enough. Here are the proposed steps:

1. Use ObjC NSString APIs to cut the PreEdit text into (maximum) 3 componenet strings using the given NSRange parameters.

2. Use the cut results and Rust string API to generate the new cursor position parameters for `Ime::Preedit(A, Option<(B,C)>`.

**Comment by ShikiSuen** [comment]
I just wrote a utility function to convert utf16 NSRange to utf8 bounds:

```rust

// Convert given NSRange to Rust String Range data used in this crate.
unsafe fn get_actual_range_bounds(source: &NSString, range: NSRange) -> Option<(usize, usize)> {
    let source_bounds = NSRange::new(0, source.length());
    let lowerbound_u16: usize = range.location;
    let upperbound_u16: usize = range.length + lowerbound_u16;
    // Sanity Check.
    let mut should_return_nil = false;
    should_return_nil |= lowerbound_u16 > source_bounds.length;
    should_return_nil |= upperbound_u16 < lowerbound_u16;
    should_return_nil |= upperbound_u16 > source_bounds.length;
    if should_return_nil { return None }
    let u16_range_from_zero_a: NSRange = NSRange::new(0, lowerbound_u16);
    let u16_range_from_zero_b: NSRange = NSRange::new(0, upperbound_u16);
    let sub_string_a = unsafe { source.substringWithRange(u16_range_from_zero_a) };
    let sub_string_b = unsafe { source.substringWithRange(u16_range_from_zero_b) };
    let lowerbound_u8: usize = sub_string_a.to_string().len();
    let upperbound_u8: usize = sub_string_b.to_string().len();
    return Some((lowerbound_u8, upperbound_u8));
}
```

**Comment by ShikiSuen** [comment]
A PR dedicated for `_selected_range` is here:
https://github.com/rust-windowing/winit/pull/3619

TODO: We need some Nihonjin IME devs' help regarding how to use _replacement_range correctly. The "_replacement_range" is at least used in Japanese IMEs. Therefore, Nihonjin IME devs are more familiar with Japanese IMEs on mac.

**Comment by kchibisov** [comment]
>Therefore, Nihonjin IME devs are more familiar with Japanese IMEs on mac.

You mean, 上手 ones? They are quite rare, at least in project I work with. In general you can just compare to other apps when doing the same input and figure out from it.

I can also try to take a look once I have time...

**Comment by ShikiSuen** [comment]
@kchibisov 

I forgot to say:
```
TypeAlias "Nihonjin" = "Japanese People"
```

Replacement Range in Japanese IMEs are used for reconversions. For example (Google Japanese Input for macOS):

https://github.com/google/mozc/blob/641260dbe918d2f6ddb3168cd43d8664ea08ca43/src/mac/GoogleJapaneseInputController.mm#L541

**Comment by kchibisov** [comment]
Just fyi, I use mozc daily, though, maybe not that advanced yet.

Do you mean from the selected range? And what is the selected range in such case? I know conversions during the preedit, is it related to text around? Could you maybe make a gif of what you're talking about, so I can understand without looking too much into it yet.

**Comment by ShikiSuen** [comment]
@kchibisov

That's why I said we need some Japanese IME developers (better Mozc ones) to help explain how the reconversion works.

I only have experiences developing Chinese IMEs. It seems that the reconversion feature is a shortcut to remove the needs of composition buffer. This makes sense since bare kana characters are also important elements in writing Japanese texts (mixing kana and kanji). However, written Chinese never mixes Pinyin / Zhuyin (except for dedicated purposes).

**Comment by ShikiSuen** [comment]
@kchibisov Also, the reconversion feature needs a JIS keyboard. I don't have one.

JIS keyboard has a dedicated physical key for reconversion.
<img width="696" alt="image" src="https://github.com/rust-windowing/winit/assets/3164826/7744b471-1e6d-485f-add6-b293757041ef">

**Comment by kchibisov** [comment]
Ah, I think I got what you mean. I'm pretty sure it works by indicating surrounding text around, so you can not really do that without adding a new API. Such API was discussed in the past, but that's about it.

@ShikiSuen maybe you need a special key on macOS, but on linux I just have a binding to convert between　漢字 and kana. Though, I'm not sure there's a way to trigger the convert just by starting selection, at least I've never seen how to.

**Comment by ShikiSuen** [comment]
@kchibisov The `replacementRange` is used for interacting with the surrounding context.

**Comment by kchibisov** [comment]
Yeah, so you need an API to indicate such range, and then `delete` send new text, that's pretty much how it's on Wayland. So this API could be left untouched for awhile until we have a separate `Window::set_ime_surrounding_text` or similar APIs. Though, I won't stop you from drafting one.

**Comment by ShikiSuen** [comment]
@kchibisov For now we can just fix the use of `_selected_range`.

**Comment by kchibisov** [comment]
Yeah, selected range matches the `start/end` cursor APIs we have pretty much, as what I can say, so shouldn't be that hard, I guess.
